programs/pluto/ikev2_ts.c

```diff
@@ -320,6 +379,111 @@ stf_status v2_emit_ts_payloads(const struct child_sa *child,
+ * @param[out]		ts		Incoming Traffic Selector substructure
+ * 					currently being processed in the TS Payload.
+ * @param[in,out]	ins		Input packet byte stream for the incoming
+ * 					TS Payload.
+ * @param[out]		seclabel_index	Set to `ts_index` if the TS_SECLABEL substructure
+ * 					parsing is successful.
+ * @param[in]		ts_index	Index of the incoming Traffic Selector substructure
+ * 					currently being processed in the TS Payload.
+ *
+ * @return	True if parsing a TS_SECLABEL substructure is successful.
+ */
+static bool parse_ts_seclabel(uint32_t* const seclabel_count,
+			      struct traffic_selector *const ts,
+			      struct pbs_in *const ins,
+			      int *const seclabel_index,
+			      int const ts_index) {
+	if (*seclabel_count == 1) {
+		/*
+		 * Current implementation only allows 1 TS_SECLABEL
+		 * substructure at most in a Traffic Selector (TS) Payload.
+		 */
+		loglog(RC_LOG_SERIOUS, "ERROR: Multiple TS_SECLABEL not supported - already found a TS_SECLABEL during input processing");
+		return false;
+	}
+
+	/* ts_seclabel: TS_SECLABEL substructure of the TS Payload. */
+	struct ikev2_ts_seclabel ts_seclabel;
+
+	/* Parse the header of the TS_SECLABEL substructure. */
+	if (!in_struct(&ts_seclabel, &ikev2_ts_seclabel_desc, ins, NULL)) {
+		loglog(RC_LOG_SERIOUS, "ERROR: Could not parse header of TS_SECLABEL substructure");
+		return false;
+	}
+	passert(ts_seclabel.isa_tssec_type == IKEv2_TS_SECLABEL);
+
+	/* TS_SECLABEL substructure payload length */
+	uint16_t const payloadLen = ts_seclabel.isa_tssec_sellen;
+	/* TS_SECLABEL substructure payload header length */
+	uint16_t const headerLen = 4;
+	if (payloadLen < headerLen) {
+		/* Bad payload length: too small */
+		loglog(RC_LOG_SERIOUS, "ERROR: TS_SECLABEL payloadLen (%u) < headerLen (%u)", payloadLen, headerLen);
+		return false;
+	}
+
+	/* Length of the security label. */
+	uint16_t const labelLen = payloadLen - headerLen;
+	if (labelLen == 0) {
+		/*
+		 * Zero-length security labels disallowed by
+		 * the labeled IPsec standard.
+		 */
+		loglog(RC_LOG_SERIOUS, "ERROR: Zero TS_SECLABEL label length");
+		return false;
+	}
+
+	if (labelLen > MAX_SECCTX_LEN) {
+		/*
+		 * Security label exceeds maximum allowed size.
+		 */
+		loglog(RC_LOG_SERIOUS, "ERROR: Incoming TS_SECLABEL label length (%u) exceeds maximum allowed buffer size (%u)", labelLen, MAX_SECCTX_LEN);
+		return false;
+	}
+
+
+	ts->ts_type = IKEv2_TS_SECLABEL;
+
+	struct xfrm_user_sec_ctx_ike incoming_ctx;
+	/* FIXME: These are hardcoded to correspond to SELinux. */
+	incoming_ctx.ctx.ctx_doi = 1;
+	incoming_ctx.ctx.ctx_alg = 1;
+
+	incoming_ctx.ctx.ctx_len = labelLen;
+
+	/* Parse the security label value of the TS_SECLABEL substructure. */
+	if (!in_raw(incoming_ctx.sec_ctx_value, labelLen, ins,
+				"Traffic Selector Security Label")) {
+		loglog(RC_LOG_SERIOUS, "ERROR: Could not read TS_SECLABEL security label value");
+		return false;
+	}
+
+	/* Incoming security label MUST be NUL-terminated. */
```

Comment:
NULL

Reply:
Same as above regarding the NUL character. But let me know if you would like me to change it to NULL still though.